src/common/range.rs

```diff
@@ -141,278 +146,454 @@ impl ByteRangeSpec {
         }
     }
 }
 
 impl FromStr for ByteRangeSpec {
-    type Err = ::Error;
+    type Err = InvalidRange;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut iter = s.trim().splitn(2, '-');
+        Ok(
+            match (parse_bound(iter.next())?, parse_bound(iter.next())?) {
+                (None, Some(len)) if len != 0 => ByteRangeSpec::Last(NonZeroU64::new(len).unwrap()),
+                (Some(first), None) => ByteRangeSpec::AllFrom(first),
+                (Some(first), Some(last)) if first <= last => ByteRangeSpec::FromTo(first, last),
+                _ => return Err(err()),
+            },
+        )
+    }
+}
+error_type!(InvalidRange);
 
-    fn from_str(s: &str) -> ::Result<ByteRangeSpec> {
-        let mut parts = s.splitn(2, '-');
+fn err() -> InvalidRange {
+    InvalidRange { _inner: () }
+}
 
-        match (parts.next(), parts.next()) {
-            (Some(""), Some(end)) => {
-                end.parse().or(Err(::Error::Header)).map(ByteRangeSpec::Last)
-            },
-            (Some(start), Some("")) => {
-                start.parse().or(Err(::Error::Header)).map(ByteRangeSpec::AllFrom)
-            },
-            (Some(start), Some(end)) => {
-                match (start.parse(), end.parse()) {
-                    (Ok(start), Ok(end)) if start <= end => Ok(ByteRangeSpec::FromTo(start, end)),
-                    _ => Err(::Error::Header)
-                }
-            },
-            _ => Err(::Error::Header)
+fn reduce_bound(bound: Bound<&u64>, start: bool) -> Option<u64> {
+    match bound {
+        Bound::Included(n) => Some(*n),
+        Bound::Excluded(n) => {
+            if start {
+                Some(n + 1)
+            } else {
+                Some(n - 1)
+            }
+        }
+        Bound::Unbounded => {
+            if start {
+                Some(0)
+            } else {
+                None
+            }
         }
     }
 }
 
-fn from_comma_delimited<T: FromStr>(s: &str) -> Vec<T> {
-    s.split(',')
-        .filter_map(|x| match x.trim() {
-            "" => None,
-            y => Some(y)
-        })
-        .filter_map(|x| x.parse().ok())
-        .collect()
+fn reduce_bounds(bounds: impl RangeBounds<u64>) -> (u64, Option<u64>) {
+    (
+        reduce_bound(bounds.start_bound(), true).unwrap(),
+        reduce_bound(bounds.end_bound(), false),
+    )
 }
 
-impl Header for Range {
-
-    fn header_name() -> &'static str {
-        static NAME: &'static str = "Range";
-        NAME
+fn parse_bound(bound: Option<&str>) -> Result<Option<u64>, InvalidRange> {
+    let mut invalid = false;
+    let bound = bound
+        .and_then(|bound| if bound.is_empty() { None } else { Some(bound) })
+        .and_then(|bound| match bound.parse() {
+            Ok(bound) => Some(bound),
+            Err(_) => {
+                invalid = true;
+                None
+            }
+        });
+    if invalid {
+        Err(err())
+    } else {
+        Ok(bound)
     }
```

Comment:
This could just be `bound.map(|bound| bound.parse()).ok_or_else(err)` I think?

Reply:
not quite, because we need differentiate between three cases basically:
 - empty string, which results in `Ok(None)`
 - some valid integer, encoded as a string, resulting in `Ok(Some(int))`
 - invalid input, which results in `Err(_)`
 
 Is this clearer? It takes out the temporary mutable boolean, so that looks a bit nicer to me, and it means that we explicitly have the three options listed above all present in the second `and_then`
 
 ```diff
 diff --git a/src/common/range.rs b/src/common/range.rs
index 7e88bb9..ba1b315 100644
--- a/src/common/range.rs
+++ b/src/common/range.rs
@@ -362,21 +362,18 @@ fn reduce_bounds(bounds: impl RangeBounds<u64>) -> (u64, Option<u64>) {
 }
 
 fn parse_bound(bound: Option<&str>) -> Result<Option<u64>, InvalidRange> {
-    let mut invalid = false;
-    let bound = bound
-        .and_then(|bound| if bound.is_empty() { None } else { Some(bound) })
-        .and_then(|bound| match bound.parse() {
-            Ok(bound) => Some(bound),
-            Err(_) => {
-                invalid = true;
-                None
+    Ok(bound.and_then(|bound| if bound.is_empty() { None } else { Some(bound) })).and_then(
+        |bound| {
+            if let Some(bound) = bound {
+                return match bound.parse() {
+                    Ok(bound) => Ok(Some(bound)),
+                    Err(_) => Err(err()),
+                };
+            } else {
+                Ok(None)
             }
-        });
-    if invalid {
-        Err(err())
-    } else {
-        Ok(bound)
-    }
+        },
+    )
 }
 ```

Reply:
Ah I see, didn't look closely enough at how this function was being called. One thing is that, if I'm not mistaken, passing `None` to this will *always* translate to an error outside of this function. With that in mind this could probably look more like

```rust
fn parse_bound(bound: &str) -> Result<Option<u64>, InvalidRange> {
    if bound.is_empty() {
        return Ok(None);
    }
    bound.parse().map_err(|_| err()) // or do something more helpful than `err`
}
```